be/src/runtime/plugin/s3_plugin_downloader.cpp

```diff
@@ -0,0 +1,185 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "runtime/plugin/s3_plugin_downloader.h"
+
+#include <fmt/format.h>
+
+#include <filesystem>
+#include <fstream>
+#include <memory>
+#include <string>
+#include <thread>
+
+#include "common/logging.h"
+#include "common/status.h"
+#include "io/fs/local_file_system.h"
+#include "io/fs/s3_file_system.h"
+#include "runtime/plugin/plugin_file_cache.h"
+#include "util/s3_util.h"
+
+namespace doris {
+
+std::string S3PluginDownloader::S3Config::to_string() const {
+    return fmt::format("S3Config{{endpoint='{}', region='{}', bucket='{}', access_key='{}'}}",
+                       endpoint, region, bucket, access_key.empty() ? "null" : "***");
+}
+
+S3PluginDownloader::S3PluginDownloader(const S3Config& config) : config_(config) {
+    s3_fs_ = create_s3_filesystem(config_);
+    // Note: We don't throw here. The caller should check via download_file return status
+}
+
+S3PluginDownloader::~S3PluginDownloader() = default;
+
+Status S3PluginDownloader::download_file(const std::string& remote_s3_path,
+                                         const std::string& local_target_path,
+                                         std::string* local_path, const std::string& expected_md5) {
+    // Check if S3 filesystem is initialized
+    if (!s3_fs_) {
+        return Status::InternalError("S3 filesystem not initialized");
+    }
+
+    // Check if download is needed first
+    if (PluginFileCache::is_file_valid(local_target_path, expected_md5)) {
+        *local_path = local_target_path;
+        return Status::OK(); // Local file is valid, return directly
+    }
+
+    // Execute download retry logic
+    Status last_status;
+    for (int attempt = 1; attempt <= MAX_RETRY_ATTEMPTS; ++attempt) {
+        last_status = execute_download(remote_s3_path, local_target_path, expected_md5);
+        if (last_status.ok()) {
+            *local_path = local_target_path;
+            return Status::OK();
+        }
+        if (attempt < MAX_RETRY_ATTEMPTS) {
```

Comment:
I don't think we need this retry logic, just let the query to retry is enough.
Otherwise, if it is network issue or the credential info is incorrect, the retry is wasteful.